### PR TITLE
overlord, daemon: add collect attribute hooks

### PR DESCRIPTION
**Pull request by stolowski**
Added two hooks for collecting plug and slot attributes via snapctl (keys+values; to be implemented separately when snapctl is ready).

### CODE REVIEW

File: overlord/ifacestate/hook_handlers.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// SetupHooks sets hooks of InterfaceManager up
+func SetupHooks(hookMgr *hookstate.HookManager) {
+	generator := func(context *hookstate.Context) hookstate.Handler {
+		return &collectAttrHandler{}
+	}
+
+	hookMgr.Register(regexp.MustCompile("^collect-plug-attr-[a-zA-Z0-9_\\-]+$"), generator)
+	hookMgr.Register(regexp.MustCompile("^collect-slot-attr-[a-zA-Z0-9_\\-]+$"), generator)
```
Comment (kyrofa) [inline overlord/ifacestate/hook_handlers.go]: If you're going to use the same generator, why not use a regex more like `^collect-(?:plug|slot)-attr-[a-zA-Z0-9_\\-]+$` ?

File: overlord/ifacestate/ifacemgr.go [hunk 2]
```diff
@@ -71,12 +73,21 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
 	// TODO: Store the intent-to-connect in the state so that we automatically
 	// try to reconnect on reboot (reconnection can fail or can connect with
 	// different parameters so we cannot store the actual connection details).
-	summary := fmt.Sprintf(i18n.G("Connect %s:%s to %s:%s"),
+	summary := fmt.Sprintf(i18n.G("Collect attributes of plug %s:%s"), plugSnap, plugName)
+	collectPlugAttr := hookstate.HookTask(s, summary, plugSnap, snap.Revision{N: 0}, "collect-plug-attr-"+plugName)
```
Comment (kyrofa) [inline overlord/ifacestate/ifacemgr.go]: `0` is the default value for an int, so you don't need to include it when creating the revision: `snap.Revision{}`

File: overlord/ifacestate/ifacemgr.go [hunk 3]
```diff
@@ -71,12 +73,21 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
...
+
+	summary = fmt.Sprintf(i18n.G("Collect attributes of slot %s:%s"), slotSnap, slotName)
+	collectSlotAttr := hookstate.HookTask(s, summary, slotSnap, snap.Revision{N: 0}, "collect-slot-attr-"+slotName)
```
Comment (kyrofa) [inline overlord/ifacestate/ifacemgr.go]: Here as well.

File: overlord/ifacestate/ifacemgr_test.go [hunk 4]
```diff
@@ -100,15 +121,18 @@ func (s *interfaceManagerSuite) TestConnectTask(c *C) {
 }
 
 func (s *interfaceManagerSuite) TestEnsureProcessesConnectTask(c *C) {
+	s.command = testutil.MockCommand(c, "snap", "")
```
Comment (kyrofa) [inline overlord/ifacestate/ifacemgr_test.go]: Where is this being restored?

File: snap/hooktypes.go [hunk 5]
```diff
@@ -26,6 +26,8 @@ import (
 var supportedHooks = []*HookType{
 	newHookType(regexp.MustCompile("^apply-config$")),
 	newHookType(regexp.MustCompile("^check-config$")),
+	newHookType(regexp.MustCompile("^collect-plug-attr-[a-zA-Z0-9_\\-]+$")),
+	newHookType(regexp.MustCompile("^collect-slot-attr-[a-zA-Z0-9_\\-]+$")),
```
Comment (kyrofa) [inline snap/hooktypes.go]: You could combine these expressions as well.

File: overlord/ifacestate/hook_handlers.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+package ifacestate
+
+import (
+	"regexp"
+
+	"github.com/snapcore/snapd/overlord/hookstate"
+)
+
+type collectAttrHandler struct {
+}
```
Comment (zyga) [inline overlord/ifacestate/hook_handlers.go]: Move `}` to the same line as `}` please.

File: daemon/api_test.go [hunk 7]
```diff
@@ -2431,6 +2435,7 @@ func (s *apiSuite) TestInterfaces(c *check.C) {
 
 func (s *apiSuite) TestConnectPlugSuccess(c *check.C) {
 	d := s.daemon(c)
+	s.command = testutil.MockCommand(c, "snap", "")
```
Comment (niemeyer) [inline daemon/api_test.go]: It'd be nicer to keep this inline as usual and as a local var:
```
command := tesutil.MockCommand(c, "snap", "")
defer command.Restore()
```
Same for all cases below.

File: overlord/ifacestate/hook_handlers.go (hunk 1 above)
Comment (niemeyer) [inline overlord/ifacestate/hook_handlers.go]: I suspect the generators will end up differently, as they'll have different behavior and different handlers.

Can we please use the following terminology for these hooks:
```
<slot name>-slot-connecting
<slot name>-slot-connected
<plug name>-plug-connecting
<plug name>-plug-connected
```

File: overlord/ifacestate/ifacemgr.go [hunk 8]
```diff
@@ -71,12 +73,21 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
 	// TODO: Store the intent-to-connect in the state so that we automatically
 	// try to reconnect on reboot (reconnection can fail or can connect with
 	// different parameters so we cannot store the actual connection details).
-	summary := fmt.Sprintf(i18n.G("Connect %s:%s to %s:%s"),
+	summary := fmt.Sprintf(i18n.G("Collect attributes of plug %s:%s"), plugSnap, plugName)
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Similarly:
```
"Prepare connection of plug %s:%s"
"Prepare connection of slot %s:%s"
"Confirm connection of plug %s:%s"
"Confirm connection of slot %s:%s"
```

File: overlord/ifacestate/ifacemgr.go (hunk 2 above)
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Indeed.

File: overlord/ifacestate/ifacemgr_test.go [hunk 9]
```diff
@@ -26,21 +26,25 @@ import (
...
 func TestInterfaceManager(t *testing.T) { TestingT(t) }
 
 type interfaceManagerSuite struct {
 	state           *state.State
 	privateMgr      *ifacestate.InterfaceManager
+	privateHookMgr  *hookstate.HookManager
 	extraIfaces     []interfaces.Interface
 	secBackend      *interfaces.TestSecurityBackend
 	restoreBackends func()
+	command         *testutil.MockCmd
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr_test.go]: Same as previous comments about command in the suite.

File: overlord/overlord.go [hunk 10]
```diff
@@ -114,12 +121,7 @@ func New() (*Overlord, error) {
...
 
-	hookMgr, err := hookstate.Manager(s)
-	if err != nil {
-		return nil, err
-	}
-	o.hookMgr = hookMgr
-	o.stateEng.AddManager(o.hookMgr)
+	ifacestate.SetupHooks(hookMgr)
 
 	configMgr, err := configstate.Manager(s, hookMgr)
```
Comment (niemeyer) [inline overlord/overlord.go]: Let's please use the same convention we've used here to pass the hook manager into the interface manager.

File: snap/hooktypes.go (hunk 5 above)
Comment (niemeyer) [inline snap/hooktypes.go]: Indeed.

File: overlord/ifacestate/hook_handlers.go (hunk 1 above)
Comment (niemeyer) [inline overlord/ifacestate/hook_handlers.go]: (update the plug/slot hook names)

File: overlord/ifacestate/ifacemgr.go (hunk 8 above)
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: (updated wording)

**Comment by niemeyer** [comment]
Conflicts and open points for a week. Please reopen when it's ready for another look.

**Pull request closed without merging**